### PR TITLE
Fix: run effect sync

### LINKED ISSUE #48: Issues with nullified events
I am experiencing an issue when directly passing an effect to an element as an event handler.

It is to do with the empty `https://github.com/FormidableLabs/freactal/blob/master/src/effects.js#L9` causing the `effectFn` to be defered to the next tick, after which React has nullified the event, making it impossible to extract the values purely via effects.

Changing:
```js
Promise.resolve()
      .then(() => effectFn(effects, ...args))
```
to:
```js
Promise.resolve(effectFn(effects, ...args))
```
Fixes the issue by calling `effectFn` immediately. I am not sure if this will negatively affect the timing of anything else.

I can see in the testing example in the readme (https://github.com/FormidableLabs/freactal/blob/master/README.md#L640) that a wrapper function is used in the render to immediately extract the value from the element.
This works fine, however I have always avoided function declaration within render for performance reasons.

Feel free to close this if it isn't considered an issue, however it is something that I have run into.

### PR DESCRIPTION
**Pull request by julien-f**
Fixes #48

Change done quickly in the browser, please test it thoroughly :)

-----

The maintainers of this repo require that all pull request submitters adhere to the following:

- [x] I have read and will comply with the [contribution guidelines](https://github.com/FormidableLabs/freactal/blob/master/CONTRIBUTE.md).
 _(required)_
- [x] I have read and will comply with the [code of conduct](https://github.com/FormidableLabs/freactal/blob/master/CONTRIBUTE.md).
 _(required)_
- [x] All related documentation has been updated to reflect the changes made. _(required)_
- [x] My commit messages are cleaned up and ready to merge. _(required)_

The maintainers of this repository require you to select the semantic version type that
the changes in this pull request represent.  Please select one of the following:
- [x]  major
- [ ]  minor
- [ ]  patch
- [ ]  documentation only

### CODE REVIEW

**Comment by julien-f** [comment]
Running the effect synchronously is necessary when handling events:

- `event.preventDefault()`
- `event.target.value`: even persisting the event is not enough because the value of the target will be reset (by React) before accessing it

File: src/effects.js [hunk 1]
```diff
@@ -11,9 +11,15 @@ export const getEffects = (hocState, effectDefs, parentEffects) => {
...
-      .then(applyReducer);
+    memo[effectKey] = (...args) => {
+      let promise;
+      try {
+        promise = Promise.resolve(effectFn(effects, ...args));
+      } catch (error) {
+        promise = Promise.reject(error);
+      }
+      return promise.then(applyReducer);
+    };
```
Comment (julien-f) [inline src/effects.js]: I can change for a more compact version if you prefer:

```js
memo[effectKey] = (...args) => new Promise(resolve =>
  resolve(effectFn(effects, ...args))
).then(applyReducer)
```

**Comment by Jascha-Sundaresan** [comment]
@divmain mentions in #48 that this "would result in an actual throw, vs a rejected promise - and, in turn, that could complicate effect composition somewhat."

is there a reason you didn't go with his proposed solution of
```
memo[effectKey] = (...args) => new Promise((resolve, reject) => {
  try {
    resolve(effectFn(effects, ...args));
  } catch (err) {
    reject(err);
  }
})
  .then(() => effectFn(effects, ...args))
  .then(applyReducer);
```
instead?

**Comment by julien-f** [comment]
Hey @Jascha-Sundaresan,

That's not necessary, `Promise.resolve(fn())` can indeed throw before being able to return a promise but that's not the case for `Promise(resolve => resolve(fn())` 😄

[Example](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=GYVwdgxgLglg9mABMJAKAlIg3gKEYqACwCc4B3RMAUwoFFjTiMcBfHHKYgT2z0QAVSAWxgBnKgDpiVUXAA2ANyqoUGdBIgBDKBEKoMiALwA-XvnwQEsuZLlwA5qgDk0gFZVoVACZP0fFn4siFo6hIioVAxwxJi4Flbytg7ORKRkYL6s7Jw8cZQ0AsJiytLWSkampfJKKmBq6iG6-pgmZuaWYNZJji5U7p4-fvgBrMHauuGRjLF8HV0Sdj2p5BmBiEA&debug=false&circleciRepo=&evaluate=true&lineWrap=true&presets=env&targets=&version=6.26.0).

**Comment by julien-f** [comment]
I there anything I can do to help this code to be merged?

**Comment by ryan-roemer** [comment]
@julien-f -- @divmain and I chatted about this on Friday and he is looking into it / the best solution (which I think may still involve the `try/catch` variant so that the end users don't have to try/catch wrap calls). At any rate, I'll leave it to Dale from here...

**Comment by julien-f** [comment]
@ryan-roemer As I explained in a previous comment, the `try`/`catch` variant does not do anything more, `new Promise(resolve => resolve(fn()))` already catch any exceptions thrown by `fn` and return a rejected promise 😄

**Comment by ryan-roemer** [comment]
Oh, haha, yeah that's totally right.

**Comment by Jascha-Sundaresan** [comment]
any word on this @divmain?

**Comment by julien-f** [comment]
Friendly poke ;)

**Comment by stefvhuynh** [comment]
@ryan-roemer, @divmain, this LGTM.

**Comment by stefvhuynh** [comment]
@divmain, anything you want to add?

**Comment by ryan-roemer** [comment]
maintainerd seems to be on the fritz again. I'm going to close and immediately re-open this PR to see if that kicks it.

**Comment by ryan-roemer** [comment]
@julien-f -- `maintainerd` has failed because you need to format the original message opening this PR specifically. (Faiure is `body - 4 checklist items remaining.`)

Can you look at closed PRs like: https://github.com/FormidableLabs/freactal/pull/76 as there's some mandatory stuff there (like semver change etc) that enable this PR to go straight to release.

Thanks!

/cc @stefvhuynh @divmain

**Comment by ryan-roemer** [comment]
Also, looks like PR needs to catch up with `master` (?)

**Comment by divmain** [comment]
Hi @julien-f.  Sorry to leave this idle.  I'm catching up now on long-standing issues.

The solution you've proposed here doesn't address my concern from the issue.  Errors will be swallowed into the abyss and not propagated down the promise chain (bad for effect composition) and potentially not reported in the console (bad generally).  See the example here:

https://codepen.io/anon/pen/POddpy?editors=1111

The suggestion I made in the issue would suffice, and I may end up making that change myself.  The other option would be a helper function that clones the event into a pseudo-event object that could be passed around asynchronously.

If it were just me, I would probably go in that direction, but I imagine others would prefer the synchronous one.  So I think we should go with that.

**Comment by ryan-roemer** [comment]
@divmain -- Don't you need

```js
resolve(throwErr())
```

instead of

```js
resolve(throwErr)
```

in your example to be equivalent to `.then(throwErr)`? That does seem to be caught...

**Comment by divmain** [comment]
You're right!  I'm basing this all on an old revision of the Promise spec, apparently.  This being the case, I'm perfectly happy to merge this as-is.

**Comment by divmain** [comment]
This is technically a breaking change.  @julien-f, I'll include this commit along with a couple of other breaking changes.  However, please do indicate that you've accepted the contribution guidelines and code of conduct.  Thanks, and sorry this has taken so long!

**Comment by julien-f** [comment]
@divmain I've updated the PR message 😄 

It's indeed technically a breaking change but I don't really see how it might break anything in real life 😛

**Comment by divmain** [comment]
@julien-f, work on the other breaking changes got pushed back until January, so I'm merging this in now.  Thanks again!

**Comment by divmain** [comment]
Published in v2.0.0.

**Pull request merged**